server/src/smart-scheduling-routes.ts

```diff
@@ -218,46 +313,70 @@ export async function listSlots(req: Request, res: Response): Promise<void> {
-        // "unknown" status. Instead, "unknown" with status = "free" and
-        // capacity = 0.
-        let capacity = 0;
-        let status = "free";
-        if (provider.availability?.available === Availability.NO) {
-          status = "busy";
-        } else if (provider.availability?.available === Availability.YES) {
-          capacity = 1;
-        }
+      .flatMap((provider) => {
+        const slots =
+          provider.availability?.slots || provider.availability?.capacity;
+        if (!slots || !slots.length) return [];
 
-        const extension: Array<any> = [
-          {
-            url: SLOT_CAPACITY,
-            valueInteger: capacity,
-          },
-        ];
-        if (provider.availability?.valid_at) {
-          extension.push({
-            url: CURRENT_AS_OF,
-            valueInstant: provider.availability.valid_at,
-          });
-        }
+        return slots.map((slot: SlotRecord | CapacityRecord) => {
+          const extension: Array<any> = [];
 
-        return JSON.stringify({
-          resourceType: "Slot",
-          // NOTE: high likelihood of going over the max length here.
-          id: `${provider.id}__covid19vaccine_combined_slot`,
-          schedule: {
-            reference: `Schedule/${provider.id}__covid19vaccine`,
-          },
-          status,
-          // These times are a lie. For most providers, we have no detailed
-          // slot information beyond "yes/no/unknown appointments available at
-          // some future time".
-          // TODO: provide more detail here when we have it.
-          start: new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString(),
-          end: new Date(Date.now() + 32 * 60 * 60 * 1000).toISOString(),
-          extension,
+          if (slot.available_count != null) {
+            extension.push({
+              url: EXTENSION_SLOT_CAPACITY,
+              valueInteger: slot.available_count,
+            });
+          }
+          if (provider.booking_phone) {
+            extension.push({
+              url: BOOKING_PHONE,
+              valueString: provider.booking_phone,
+            });
+          }
+          if (provider.booking_url) {
+            extension.push({
+              url: BOOKING_DEEP_LINK,
+              valueUrl: provider.booking_url,
+            });
+          }
+
+          let id = provider.id;
+          let start, end;
+          if ("start" in slot) {
+            start = slot.start;
+            // FHIR slots require an end time. Scrapers don’t always show one,
+            // so assume 15 minute slots when unknown.
+            end =
+              slot.end ||
+              new Date(
+                new Date(slot.start).getTime() + 15 * 60 * 1000
+              ).toISOString();
+            id = `${id}-${new Date(slot.start).getTime()}`;
+          } else {
+            start = `${slot.date}T00:00:00Z`;
+            end = `${slot.date}T00:00:00Z`;
+            id = `${id}-${slot.date}`;
```

Comment:
Not sure whether it would buy you much here, but it might be good to pull out the time formatting stuff here into functions (or rely on `luxon`?)

Reply:
Yeah, I guess I really should have done that for the math above. Not sure Luxon is super valuable to these particular lines since they’re so simple, but I hear you on cleaning this up a bit.